projects/plugins/jetpack/modules/sso.php

```diff
@@ -81,6 +82,121 @@ private function __construct() {
 			require_once JETPACK__PLUGIN_DIR . 'modules/sso/class-jetpack-force-2fa.php';
 			new Jetpack_Force_2FA();
 		}
+		add_filter( 'manage_users_columns', array( $this, 'jetpack_user_connected_th' ) );
+		add_action( 'admin_print_styles-users.php', array( $this, 'jetpack_user_table_styles' ) );
+		add_action( 'manage_users_custom_column', array( $this, 'jetpack_show_connection_status' ), 10, 3 );
+		add_action( 'admin_post_jetpack_invite_user_to_wpcom', array( $this, 'invite_user_to_wpcom' ) );
+		add_action( 'admin_notices', array( $this, 'handle_invitation_results' ) );
+	}
+
+	/**
+	 * Renders invitations errors/success messages in users.php.
+	 */
+	public function handle_invitation_results() {
+		$valid_nonce = isset( $_GET['_wpnonce'] ) ? wp_verify_nonce( $_GET['_wpnonce'], 'jetpack-sso-invite-user' ) : false; // phpcs:ignore WordPress.Security.ValidatedSanitizedInput -- WP core doesn't pre-sanitize nonces either.
+
+		if ( ! isset( $_GET['jetpack-sso-invite-user'] ) || ! $valid_nonce ) {
+			return;
+		}
+		if ( $_GET['jetpack-sso-invite-user'] === 'success' ) {
+			return wp_admin_notice( __( 'Hooray! User was invited successfully!', 'jetpack' ), array( 'type' => 'success' ) );
+		}
+		if ( $_GET['jetpack-sso-invite-user'] === 'failed' && isset( $_GET['jetpack-sso-invite-error'] ) ) {
+			switch ( $_GET['jetpack-sso-invite-error'] ) {
+				case 'invalid-user':
+					return wp_admin_notice( __( 'Tried to invite a user that either doesn&#8217;t exist.', 'jetpack' ), array( 'type' => 'error' ) );
+				case 'invalid-email':
+					return wp_admin_notice( __( 'Tried to invite a user that doesn&#8217;t have an email address.', 'jetpack' ), array( 'type' => 'error' ) );
+				default:
+					// render whatever error we got from the API. It's escaped and nonce-checked.
+					return wp_admin_notice( esc_html( sanitize_text_field( wp_unslash( $_GET['jetpack-sso-invite-error'] ) ) ), array( 'type' => 'error' ) );
+			}
+		}
+	}
+
+	/**
+	 * Invites a user to connect to WordPress.com to allow them to log in via SSO.
+	 */
+	public function invite_user_to_wpcom() {
+		check_ajax_referer( 'jetpack-sso-invite-user', 'invite_nonce' );
+		$nonce = wp_create_nonce( 'jetpack-sso-invite-user' );
+
+		if ( isset( $_GET['user_id'] ) ) {
```

Comment:
Before inviting a user we should make sure the person making that change has the necessary capabilities to do so and check for `create_users`.

Reply:
I agree but it's probably `promote_user`. That's what the wpcom API requires.

Reply:
That link doesn't seem to be working? My understanding is that promote_users is for role changes and to add users in a multisite network

Reply:
Here's an alternative: 334f8-pb

I agree with the `create_users` permission, but it will be inconsistent with Calypso and users will find they can add users in Calypso but not wp-admin.

Reply:
Thanks for the context! I guess if we want to be super correct about it in the endpoint, we should check `promote_users` on simple sites and `create_users` on WoA sites ([reference](https://github.com/WordPress/WordPress/blob/cdbe6c71a5cd08d8a5b68f6509b19baf98d5fc5e/wp-includes/admin-bar.php#L981)), assuming that works and users don't have a different set of capabilities in the context of API requests.

I could also see us using `promote_users` here from the point of view that we're adding a WP.com user.